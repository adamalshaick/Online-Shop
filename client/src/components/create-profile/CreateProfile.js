import React, { Component } from "react";
import { connect } from "react-redux";
import { withRouter } from "react-router-dom";
import PropTypes from "prop-types";
import TextFieldGroup from "../common/TextFieldGroup";
import TextAreaFieldGroup from "../common/TextAreaFieldGroup";
import InputGroup from "../common/InputGroup";
import { createProfile } from "../../actions/profileActions";

export class CreateProfile extends Component {
  constructor(props) {
    super(props);
    this.state = {
      displaySocialInputs: false,
      handle: "",
      location: "",
      bio: "",
      twitter: "",
      facebook: "",
      linkedin: "",
      youtube: "",
      instagram: "",
      errors: {}
    };
  }
  componentWillReceiveProps(nextProps) {
    if (nextProps.errors) {
      this.setState({ errors: nextProps.errors });
    }
  }

  onSubmit = e => {
    e.preventDefault();

    const profileData = {
      handle: this.state.handle,
      location: this.state.location,
      bio: this.state.bio,
      twitter: this.state.twitter,
      facebook: this.state.facebook,
      linkedin: this.state.linkedin,
      youtube: this.state.youtube,
      instagram: this.state.instagram
    };

    this.props.createProfile(profileData, this.props.history);
  };

  onChange = e => {
    this.setState({ [e.target.name]: e.target.value });
  };

  render() {
    const { errors, displaySocialInputs } = this.state;

    let socialInputs;

    if (displaySocialInputs) {
      socialInputs = (
        <div>
          <InputGroup
            id="twitter"
            placeholder="Twitter Profile URL"
            name="twitter"
            icon="fab fa-twitter"
            value={this.state.twitter}
            onChange={this.onChange}
            error={errors.twitter}
          />

          <InputGroup
            id="facebook"
            placeholder="Facebook Page URL"
            name="facebook"
            icon="fab fa-facebook"
            value={this.state.facebook}
            onChange={this.onChange}
            error={errors.facebook}
          />

          <InputGroup
            id="linkedin"
            placeholder="Linkedin Profile URL"
            name="linkedin"
            icon="fab fa-linkedin"
            value={this.state.linkedin}
            onChange={this.onChange}
            error={errors.linkedin}
          />

          <InputGroup
            id="youtube"
            placeholder="YouTube Channel URL"
            name="youtube"
            icon="fab fa-youtube"
            value={this.state.youtube}
            onChange={this.onChange}
            error={errors.youtube}
          />

          <InputGroup
            id="instagram"
            placeholder="Instagram Page URL"
            name="instagram"
            icon="fab fa-instagram"
            value={this.state.instagram}
            onChange={this.onChange}
            error={errors.instagram}
          />
        </div>
      );
    }

    return (
      <div className="create-profile entry mt-5">
        <div className="row">
          <div
            style={{
              boxShadow:
                "0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)",
              padding: "2rem"
            }}
            className="col-md-8 m-auto"
          >
            <h1 style={{ fontSize: "1.7rem" }} className="mb-4 text-center">
              Create Your Profile
            </h1>

            <small className="d-block pb-3">* = required fields</small>
            <form onSubmit={this.onSubmit}>
              <TextFieldGroup
                id="handle"
                placeholder="* Profile Handle"
                name="handle"
                value={this.state.handle}
                onChange={this.onChange}
                error={errors.handle}
                info="Your username"
              />

              <TextFieldGroup
                id="location"
                placeholder="Location"
                name="location"
                value={this.state.location}
                onChange={this.onChange}
                error={errors.location}
                info="Your Location"
              />

              <TextAreaFieldGroup
                id="bio"
                placeholder="Short Bio"
                name="bio"
                value={this.state.bio}
                onChange={this.onChange}
                error={errors.bio}
                info="Tell us a little about yourself"
              />

              <div className="mb-3">
                <button
                  type="button"
                  onClick={() => {
                    this.setState(prevState => ({
                      displaySocialInputs: !prevState.displaySocialInputs
                    }));
                  }}
                  className="btn btn-dark"
                >
                  Add Social Media Links
                </button>
                <small className="text-muted d-block">(Optional)</small>
              </div>
              {socialInputs}
              <input
                type="submit"
                value="Submit"
                className="btn btn-dark float-right mt-4"
              />
            </form>
          </div>
        </div>
      </div>
    );
  }
}

CreateProfile.propTypes = {
  profile: PropTypes.object.isRequired,
  errors: PropTypes.object.isRequired,
  createProfile: PropTypes.func.isRequired
};

const mapStateToProps = state => ({
  profile: state.profile,
  errors: state.errors
});

export default connect(
  mapStateToProps,
  { createProfile }
)(CreateProfile);
