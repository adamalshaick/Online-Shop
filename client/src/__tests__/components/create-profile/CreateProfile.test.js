import React from "react";
import { CreateProfile } from "../../../components/create-profile/CreateProfile";
import { shallow } from "enzyme";
import toJson from "enzyme-to-json";

const mockCreateProfilefn = jest.fn();
const Props = {
  errors: {}
};
let wrapper;

wrapper = shallow(
  <CreateProfile {...Props} createProfile={mockCreateProfilefn} />
);
describe("render component", () => {
  it("renders component", () => {
    expect(toJson(wrapper)).toMatchSnapshot();
  });
});

it("displays form", () => {
  expect(wrapper.find("form").length).toEqual(1);
});

it("should call the mock create profile function", () => {
  wrapper.find("form").simulate("submit", { preventDefault() {} });
  expect(mockCreateProfilefn.mock.calls.length).toBe(1);
});

describe("create profile action", () => {
  const profileData = new FormData();
  profileData.append("handle", "handle value");
  profileData.append("location", "location value");
  profileData.append("bio", "bio value");

  wrapper
    .find("#handle")
    .simulate("change", { target: { name: "handle", value: "handle value" } });
  wrapper.find("#location").simulate("change", {
    target: { name: "location", value: "location value" }
  });
  wrapper
    .find("#bio")
    .simulate("change", { target: { name: "bio", value: "bio value" } });

  it("calls function with correct data", () => {
    expect(mockCreateProfilefn.mock.calls[0][0]).toEqual(profileData);
  });
});
