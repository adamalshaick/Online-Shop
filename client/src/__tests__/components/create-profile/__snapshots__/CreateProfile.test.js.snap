// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`render component renders component 1`] = `
<Fragment>
  <Connect(Navbar) />
  <div
    className="entry mt-5 container"
  >
    <div
      className="row"
    >
      <div
        className="col-md-8 m-auto"
        style={
          Object {
            "boxShadow": "0 4px 8px 0 rgba(0, 0, 0, 0.2), 0 6px 20px 0 rgba(0, 0, 0, 0.19)",
            "padding": "2rem",
          }
        }
      >
        <h1
          className="mb-4 text-center"
          style={
            Object {
              "fontSize": "1.7rem",
            }
          }
        >
          Create Your Profile
        </h1>
        <small
          className="d-block pb-3"
        >
          * = required fields
        </small>
        <form
          onSubmit={[Function]}
        >
          <div
            className="row"
          >
            <div
              className="col-md-6 mb-2"
            >
              <UploadFileGroup
                icon="fas fa-file-upload fa-8x"
                info="Profile Image"
                name="file"
                onChange={[Function]}
                type="file"
              />
            </div>
            <div
              className="col-md-6 mb-2"
            >
              <TextFieldGroup
                id="handle"
                info="Your username"
                name="handle"
                onChange={[Function]}
                placeholder="* Profile Handle"
                type="text"
                value="handle value"
              />
              <TextFieldGroup
                id="location"
                info="Your Location"
                name="location"
                onChange={[Function]}
                placeholder="Location"
                type="text"
                value="location value"
              />
            </div>
          </div>
          <TextAreaFieldGroup
            id="bio"
            info="Tell us a little about yourself"
            name="bio"
            onChange={[Function]}
            placeholder="Short Bio"
            value="bio value"
          />
          <input
            className="btn btn-dark float-right mt-4"
            type="submit"
            value="Submit"
          />
        </form>
      </div>
    </div>
  </div>
</Fragment>
`;
